File: packages/quran-engine-js/src/audio.js
// @ts-check
/**
 * Recitation audio URL builders + reciter directory.
 *
 * Mirrors QuranPlayer.swift. Two independent feeds:
 *   • Full-surah  : `surahLink + zeroPad3(surah) + ".mp3"`   (mp3quran.net CDNs)
 *   • Ayah-by-ayah: `https://cdn.islamic.network/quran/audio/{bitrate}/{identifier}/{globalAyah}.mp3`
 */

/**
 * @typedef {Object} Reciter
 * @property {string} id              "{name}|{qiraah??'Hafs'}|{surahLink}"
 * @property {string} name
 * @property {string} ayahIdentifier  e.g. "ar.alafasy"
 * @property {string} ayahBitrate     e.g. "128" (string, used verbatim)
 * @property {string} surahLink       full-surah CDN base, trailing slash
 * @property {string|null} qiraah     null => Hafs; else riwayah label
 * @property {string} [group]
 */

/** @param {number} n */
function pad3(n) {
  return String(n).padStart(3, "0");
}

/**
 * Full-surah recitation URL.
 * @param {Reciter} reciter
 * @param {number} surahNumber 1..114
 */
export function surahAudioUrl(reciter, surahNumber) {
  if (!(surahNumber >= 1 && surahNumber <= 114)) throw new RangeError(`surah out of range: ${surahNumber}`);
  if (!reciter.surahLink) throw new Error(`Reciter "${reciter.name}" has no full-surah feed`);
  return `${reciter.surahLink}${pad3(surahNumber)}.mp3`;
}

/**
 * Ayah-by-ayah recitation URL. Requires the global ayah number (1..6236); use
 * `Quran.globalAyahNumber(surah, ayah)` to compute it.
 * @param {Reciter} reciter
 * @param {number} globalAyahNumber
 */
export function ayahAudioUrl(reciter, globalAyahNumber) {
  return `https://cdn.islamic.network/quran/audio/${reciter.ayahBitrate}/${reciter.ayahIdentifier}/${globalAyahNumber}.mp3`;
}

const MINSHAWI_FALLBACK_NAME = "Muhammad Al-Minshawi (Murattal)";

/** True if this reciter falls back to Minshawi for individual-ayah audio. Mirrors defaultToMinshawi. */
export function defaultsToMinshawi(reciter) {
  return reciter.ayahIdentifier.includes("minshawi") && !reciter.name.includes("Minshawi");
}

/** Display name to show while ayah audio plays (honest about the fallback). */
export function ayahNowPlayingName(reciter) {
  if (defaultsToMinshawi(reciter)) return MINSHAWI_FALLBACK_NAME;
  if (reciter.qiraah) return `${reciter.name} (${reciter.qiraah})`;
  return reciter.name;
}

export class Reciters {
  /** @param {Reciter[]} list parsed data/reciters.json */
  constructor(list) {
    /** @type {Reciter[]} */
    this.list = [...list].sort((a, b) => a.name.localeCompare(b.name));
    /** @type {Map<string,Reciter>} */
    this._byId = new Map(this.list.map((r) => [r.id, r]));
  }

  all() { return this.list; }

  /** @param {string} id */
  byId(id) { return this._byId.get(id); }

  /** Reciters that have a full-surah feed. */
  withSurahFeed() { return this.list.filter((r) => r.surahLink && !r.surahLink.endsWith(".mp3")); }

  /** Reciters for a given riwayah label (null/"hafs" => the default Hafs feeds). @param {string|null} qiraah */
  byQiraah(qiraah) {
    if (!qiraah || qiraah.toLowerCase() === "hafs") return this.list.filter((r) => !r.qiraah);
    return this.list.filter((r) => r.qiraah === qiraah);
  }

  /** Distinct riwayah labels available (excluding default Hafs). */
  qiraat() {
    return [...new Set(this.list.map((r) => r.qiraah).filter(Boolean))];
  }
}

File: packages/quran-engine-py/quran_engine/audio.py
"""Recitation audio URL builders + reciter directory. Mirrors src/audio.js / docs 04-05."""
from __future__ import annotations
from .models import Reciter

_MINSHAWI_FALLBACK_NAME = "Muhammad Al-Minshawi (Murattal)"


def _pad3(n: int) -> str:
    return f"{n:03d}"


def surah_audio_url(reciter: Reciter, surah_number: int) -> str:
    if not (1 <= surah_number <= 114):
        raise ValueError(f"surah out of range: {surah_number}")
    if not reciter.surah_link:
        raise ValueError(f'Reciter "{reciter.name}" has no full-surah feed')
    return f"{reciter.surah_link}{_pad3(surah_number)}.mp3"


def ayah_audio_url(reciter: Reciter, global_ayah_number: int) -> str:
    return (
        "https://cdn.islamic.network/quran/audio/"
        f"{reciter.ayah_bitrate}/{reciter.ayah_identifier}/{global_ayah_number}.mp3"
    )


def defaults_to_minshawi(reciter: Reciter) -> bool:
    return "minshawi" in reciter.ayah_identifier and "Minshawi" not in reciter.name


def ayah_now_playing_name(reciter: Reciter) -> str:
    if defaults_to_minshawi(reciter):
        return _MINSHAWI_FALLBACK_NAME
    if reciter.qiraah:
        return f"{reciter.name} ({reciter.qiraah})"
    return reciter.name


class Reciters:
    def __init__(self, reciters: list[Reciter]):
        self.list = sorted(reciters, key=lambda r: r.name)
        self._by_id = {r.id: r for r in self.list}

    def all(self) -> list[Reciter]:
        return self.list

    def by_id(self, rid: str) -> Reciter | None:
        return self._by_id.get(rid)

    def with_surah_feed(self) -> list[Reciter]:
        return [r for r in self.list if r.surah_link and not r.surah_link.endswith(".mp3")]

    def by_qiraah(self, qiraah: str | None) -> list[Reciter]:
        if not qiraah or qiraah.lower() == "hafs":
            return [r for r in self.list if not r.qiraah]
        return [r for r in self.list if r.qiraah == qiraah]

    def qiraat(self) -> list[str]:
        seen, out = set(), []
        for r in self.list:
            if r.qiraah and r.qiraah not in seen:
                seen.add(r.qiraah)
                out.append(r.qiraah)
        return out
